markets/dagstore/wrapper.go

```diff
@@ -3,89 +3,73 @@ package dagstore
 import (
 	"context"
 	"errors"
-	"io"
+	"os"
 	"sync"
 	"time"
 
-	"github.com/filecoin-project/dagstore/index"
 	"github.com/ipfs/go-cid"
 	ds "github.com/ipfs/go-datastore"
-	bstore "github.com/ipfs/go-ipfs-blockstore"
+	levelds "github.com/ipfs/go-ds-leveldb"
+	measure "github.com/ipfs/go-ds-measure"
 	logging "github.com/ipfs/go-log/v2"
+	ldbopts "github.com/syndtr/goleveldb/leveldb/opt"
 	"golang.org/x/xerrors"
 
+	"github.com/filecoin-project/lotus/node/config"
+
 	"github.com/filecoin-project/dagstore"
+	"github.com/filecoin-project/dagstore/index"
 	"github.com/filecoin-project/dagstore/mount"
 	"github.com/filecoin-project/dagstore/shard"
-	"github.com/filecoin-project/go-fil-markets/carstore"
-	"github.com/filecoin-project/go-fil-markets/shared"
+
+	"github.com/filecoin-project/go-fil-markets/stores"
 )
 
 const maxRecoverAttempts = 1
 
 var log = logging.Logger("dagstore-wrapper")
 
-// MarketDAGStoreConfig is the config the market needs to then construct a DAG Store.
-type MarketDAGStoreConfig struct {
-	TransientsDir             string
-	IndexDir                  string
-	Datastore                 ds.Datastore
-	MaxConcurrentIndex        int
-	MaxConcurrentReadyFetches int
-	GCInterval                time.Duration
-}
-
-// DAGStore provides an interface for the DAG store that can be mocked out
-// by tests
-type DAGStore interface {
-	RegisterShard(ctx context.Context, key shard.Key, mnt mount.Mount, out chan dagstore.ShardResult, opts dagstore.RegisterOpts) error
-	AcquireShard(ctx context.Context, key shard.Key, out chan dagstore.ShardResult, _ dagstore.AcquireOpts) error
-	RecoverShard(ctx context.Context, key shard.Key, out chan dagstore.ShardResult, _ dagstore.RecoverOpts) error
-	GC(ctx context.Context) (*dagstore.GCResult, error)
-	Close() error
-	Start(ctx context.Context) error
-}
-
-type closableBlockstore struct {
-	bstore.Blockstore
-	io.Closer
-}
-
 type Wrapper struct {
 	ctx          context.Context
 	cancel       context.CancelFunc
 	backgroundWg sync.WaitGroup
 
-	dagStore   DAGStore
-	mountApi   LotusAccessor
+	dagst      dagstore.Interface
+	mountApi   MinerAPI
 	failureCh  chan dagstore.ShardResult
 	traceCh    chan dagstore.Trace
 	gcInterval time.Duration
 }
 
-var _ shared.DagStoreWrapper = (*Wrapper)(nil)
+var _ stores.DAGStoreWrapper = (*Wrapper)(nil)
 
-func NewDagStoreWrapper(cfg MarketDAGStoreConfig, mountApi LotusAccessor) (*Wrapper, error) {
+func NewDAGStore(cfg config.DAGStoreConfig, mountApi MinerAPI) (*dagstore.DAGStore, *Wrapper, error) {
```

Comment:
Why do we need to return *dagstore.DAGStore here? Is it just so we can call close on the DAGStore?
If so I'd suggest adding a Close method to the Wrapper instead. That way the caller only has to deal with one interface, and we could avoid needing the DI trickery with the "special" type

Reply:
Ah no, we need `*dagstore.DAGStore` to be outjected/exported to the DI context because we'll inject it in the JSON-RPC layer to serve operations like `DagstoreShardList`, `DagstoreShardInitialize`.

Reply:
An alternative would be to _embed_ the `*dagstore.DAGStore` into the `*Wrapper`, but that is more non-obvious and requires a deeper refactor (there are name collisions).

Reply:
Oh, another thing. fx has pretty nice / first-class support for multiple outjections from a constructor: https://pkg.go.dev/go.uber.org/fx?utm_source=godoc#Out. It's just that the utils we've created on the Lotus side make this a "special" case.

Reply:
I see, makes sense 👍 

It _might_ be simpler to have a separate method that returns the `*dagstore.DAGStore`, and then inject that into DAGStoreWrapper through DI, but defer to your judgment as it sounds like you have plans for where to use the `*dagstore.DAGStore` in the near future.